// client-list/index.js
var Client = (function() {
	// Constructor
	var c = function(username, connection) {
		this.username = username;
		this.connection = connection;
		this.data = null;
	}

	return c;
})();

var ClientList = (function() {
	// Constructor
	var c = function() {
		var clients = {};
		var numberOfClients = 0;

		this.size = function() { return numberOfClients; };
	}

	c.prototype = {
		pushClient : function(client) {
			this.clients[client.username] = client;
			this.numberOfClients++;
		},

		removeClient : function(username) {
			delete clients[username];
			this.numberOfClients--;
		},

		getConnection : function(username) {
			return this.clients[username];
		},

		getDataByName : function(username, dataName) {
			if (this.clients.hasOwnProperty(username))
				if (this.clients[username].hasOwnProperty(dataName))
					return this.clients[username].data[dataName];

			return null;
		},

		forEach : function(code) {
			for (var username in this.clients)
				code.call(username);
		}
	};

	return c;
})();

module.exports = Client;
module.exports = ClientList;
